### PR TITLE
[WIP] Extend exception handler support

### LINKED ISSUE #53: Filtered exception handlers
**Summary:**
The CIL architecture supports exception handlers that are decorated with a filter block which is executed before control is transferred to the handler block. This filter block determines whether the virtual machine should enter the handler block, or pass on the exception to the upper scope. Currently, Echo has no way of representing these in terms of the `ExceptionHandlerRegion` and/or `ExceptionHandlerBlock`.

### PR DESCRIPTION
**Pull request by zsr2531**
Closes #53.

### CODE REVIEW

File: src/Core/Echo.Ast/Construction/AstParser.cs [hunk 1]
```diff
@@ -71,10 +72,26 @@ public ControlFlowGraph<Statement<TInstruction>> Parse()
...
-            // Fix entry point.
+            // Fix entry point(s).
             newGraph.Entrypoint = newGraph.Nodes[_controlFlowGraph.Entrypoint.Offset];
+            FixEntryPoint(_controlFlowGraph);
 
             return newGraph;
+
+            void FixEntryPoint(IControlFlowRegion<TInstruction> region)
+            {
+                foreach (var child in region.GetSubRegions().Where(r => r is {}))
```
Comment (Washi1337) [inline src/Core/Echo.Ast/Construction/AstParser.cs:83]: When does the subregions collection contain `null` objects?

File: src/Core/Echo.ControlFlow/Blocks/ExceptionHandlerBlock.cs [hunk 2]
```diff
@@ -16,6 +16,14 @@ public ScopeBlock<TInstruction> ProtectedBlock
             get;
         } = new ScopeBlock<TInstruction>();
 
+        /// <summary>
+        /// Gets the prologue block.
+        /// </summary>
+        public ScopeBlock<TInstruction> PrologueBlock
+        {
+            get;
+        } = new ScopeBlock<TInstruction>();
```
Comment (Washi1337) [inline src/Core/Echo.ControlFlow/Blocks/ExceptionHandlerBlock.cs:25]: Not all blocks have prologues

File: src/Core/Echo.ControlFlow/Blocks/ExceptionHandlerBlock.cs [hunk 3]
```diff
@@ -24,17 +32,31 @@ public IList<ScopeBlock<TInstruction>> HandlerBlocks
             get;
         } = new List<ScopeBlock<TInstruction>>();
 
+        /// <summary>
+        /// Gets the epilogue block.
+        /// </summary>
+        public ScopeBlock<TInstruction> EpilogueBlock
```
Comment (Washi1337) [inline src/Core/Echo.ControlFlow/Blocks/ExceptionHandlerBlock.cs:38]: Not all blocks have epilogues.

File: src/Core/Echo.ControlFlow/Regions/ExceptionHandlerRegion.cs [hunk 4]
```diff
@@ -39,6 +55,19 @@ public RegionCollection<TInstruction> HandlerRegions
...
 
+        /// <summary>
+        /// Gets the region of nodes that form the code that proceeds the handler(s).
+        /// </summary>
+        public BasicControlFlowRegion<TInstruction> EpilogueRegion
+        {
+            get => _epilogue;
+            set
+            {
+                _epilogue = value;
```
Comment (Washi1337) [inline src/Core/Echo.ControlFlow/Regions/ExceptionHandlerRegion.cs:66]: This doesn't reset the ParentRegion of the original `_epilogue` to `null`.

File: src/Core/Echo.ControlFlow/Regions/ExceptionHandlerRegion.cs [hunk 5]
```diff
@@ -31,6 +34,19 @@ public BasicControlFlowRegion<TInstruction> ProtectedRegion
...
 
+        /// <summary>
+        /// Gets the region of nodes that form the code that precedes the handler(s).
+        /// </summary>
+        public BasicControlFlowRegion<TInstruction> PrologueRegion
+        {
+            get => _prologue;
+            set
+            {
+                _prologue = value;
```
Comment (Washi1337) [inline src/Core/Echo.ControlFlow/Regions/ExceptionHandlerRegion.cs:45]: This doesn't reset the ParentRegion of the original `_prologue` to `null`.

**Review by Washi1337 — CHANGES_REQUESTED** [review]
Few things that need to be addressed.

File: src/Core/Echo.Ast/Construction/AstParser.cs (hunk 1 above)
Comment (zsr2531) [reply src/Core/Echo.Ast/Construction/AstParser.cs:83]: Forgot to remove that, my bad :)

File: src/Core/Echo.ControlFlow/Regions/HandlerRegion.cs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        /// <summary>
+        /// Gets the region of nodes that form the code that proceeds the handler.
+        /// </summary>
+        public ScopeRegion<TInstruction> Epilogue
+        {
+            get => _epilogue;
+            set => UpdateChildRegion(ref _epilogue, value);
+        }
+
+        private void UpdateChildRegion(ref ScopeRegion<TInstruction> field, ScopeRegion<TInstruction> value)
```
Comment (zsr2531) [inline src/Core/Echo.ControlFlow/Regions/HandlerRegion.cs:57]: Need to check if the `value` already belongs to another region.

**Review by zsr2531 — COMMENTED** [review]
LGTM!

**Pull request merged**